core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -272,6 +277,74 @@ public int getCapacity()
         return hashCapacity;
     }
 
+    private int putIfAbsent(int firstPosition, int positions, Page page, @Nullable BlockBuilder groupIdsBuilder)
+    {
+        /*
+         * Instead of looking for a match on a per-position basis, we execute steps in a "vectorized" way,
+         * so that all cpu mechanisms (caching, out-of-order execution, etc.) are fully utilized.
+         */
+        long[] rawHashes = new long[positions];
+        int[] hashPositions = new int[positions];
+        long[] groupAddresses = new long[positions];
+
+        for (int i = 0; i < positions; i++) {
+            rawHashes[i] = hashGenerator.hashPosition(firstPosition + i, page);
+        }
+
+        for (int i = 0; i < positions; i++) {
+            hashPositions[i] = (int) getHashPosition(rawHashes[i], mask);
+        }
+
+        for (int i = 0; i < positions; i++) {
+            groupAddresses[i] = groupAddressByHash[hashPositions[i]];
+        }
+
+        for (int i = 0; i < positions; i++) {
+            long rawHash = rawHashes[i];
+            int hashPosition = hashPositions[i];
+            long groupAddress = groupAddresses[i];
+
+            // look for an empty slot or a slot containing this key
+            int groupId = -1;
+            while (groupAddress != -1 || groupAddressByHash[hashPosition] != -1) {
```

Comment:
I have hard time comprehending this loop but it seems that we should be able to get rid of either `groupAddress` or `groupAddressByHash[hashPosition]`. And just use one of those.

Reply:
I will add a comment